### DISCUSSION TITLE
question. Reinitialization provide

### DISCUSSION THREAD

**Discussion by LazarenkoA** [uber-go/dig#357] [Q&A]
hello. I have such a code
```go
container.Provide(func(clientCfg *config.Ignite) database.Db {
		ignite, err := database.NewIgnite(ctx, clientCfg)
		if err != nil {
			logger.Error(fmt.Errorf("container provide error: %w", err))
			return nil
		} else {
			return ignite
		}
	})
```

if the database is not available at the time of initialization, further calls will return nil
```go
container.Invoke(func(ignite database.Db) error {
      //....
})
```
is it possible to make it so that when an error occurs `atabase.NewIgnite(ctx, clientCfg)` the container didn't remember the response, but the code went to `container.Provide` again?

**Comment by sywhang** [comment]
The container can't just ignore `nil` as a completely irrelevant value because nil could still be a valid value for other types. 

You could use [value groups](https://pkg.go.dev/go.uber.org/dig#hdr-Value_Groups) to do something similar to what you wanted to do. 

```go
type DBResult struct {
  dig.Out

  db database.Db `group:"db"`
}

     // somewhere in your code invokes this code many times
     container.Provide(func(clientCfg *config.Ignite) DBResult {
		ignite, err := database.NewIgnite(ctx, clientCfg)
		if err != nil {
			logger.Error(fmt.Errorf("container provide error: %w", err))
			return DBResult{
                           db: nil
                         }
		} else {
			return DBResult{ db: ignite }
		}
	})
```

You can then consume this as a list of Dbs that you can iterate and ignore nils:

```
type ServerParams struct {
  dig.In

  Dbs []database.Db `group:"db"`
}

dig.Invoke(StartServer)
func StartServer(p ServerParams) {
  // use p.Dbs to find the non-nil db.  
}
```
